frame/assets/src/migration.rs

```diff
@@ -0,0 +1,92 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2017-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// use crate::{Config, Pallet, Weight};
+// use frame_support::{
+// 	storage::migration,
+// 	traits::{Get, PalletInfoAccess},
+// };
+// use sp_std::prelude::*;
+
+use super::*;
+use frame_support::{log, traits::OnRuntimeUpgrade};
+
+pub mod v1 {
+	use frame_support::{pallet_prelude::*, weights::Weight};
+
+	use super::*;
+
+	#[derive(Decode)]
+	pub struct OldAssetDetails<Balance, AccountId, DepositBalance> {
+		pub owner: AccountId,
+		pub issuer: AccountId,
+		pub admin: AccountId,
+		pub freezer: AccountId,
+		pub supply: Balance,
+		pub deposit: DepositBalance,
+		pub min_balance: Balance,
+		pub is_sufficient: bool,
+		pub accounts: u32,
+		pub sufficients: u32,
+		pub approvals: u32,
+		pub is_frozen: bool,
+	}
+
+	impl<Balance, AccountId, DepositBalance> OldAssetDetails<Balance, AccountId, DepositBalance> {
+		fn migrate_to_v1(self) -> AssetDetails<Balance, AccountId, DepositBalance> {
+			AssetDetails {
+				owner: self.owner,
+				issuer: self.issuer,
+				admin: self.admin,
+				freezer: self.freezer,
+				supply: self.supply,
+				deposit: self.deposit,
+				min_balance: self.min_balance,
+				is_sufficient: self.is_sufficient,
+				accounts: self.accounts,
+				sufficients: self.sufficients,
+				approvals: self.approvals,
+				is_frozen: self.is_frozen,
+				status: AssetStatus::Live,
+			}
+		}
+	}
+
+	pub struct MigrateToV1<T>(sp_std::marker::PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToV1<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let current_version = Pallet::<T>::current_storage_version();
+			let onchain_version = Pallet::<T>::on_chain_storage_version();
+			if onchain_version < 1 {
+				let mut translated = 0u64;
+				Asset::<T>::translate::<
+					OldAssetDetails<T::Balance, T::AccountId, DepositBalanceOf<T>>,
+					_,
+				>(|_key, old_value| {
+					translated.saturating_inc();
+					Some(old_value.migrate_to_v1())
+				});
+				current_version.put::<Pallet<T>>();
```

Comment:
I am not sure this going to be 1
You can also have unit tests for this migration 
One example
https://github.com/paritytech/substrate/blob/70930ac26b454358c0eb1951c95d89f0920aa361/frame/alliance/src/lib.rs#L237
https://github.com/paritytech/substrate/blob/70930ac26b454358c0eb1951c95d89f0920aa361/frame/alliance/src/migration.rs#L23
please double check this

Reply:
I was following a model from a different pallet. But yeah, it might be better to be more explicit. I think i'll try the migration unit test and see. I think I couldnt find examples in the codebase

Reply:
I found the example I was following: https://github.com/paritytech/substrate/blob/aa/safely-destroy-large-assets/frame/nomination-pools/src/migration.rs#L88-L88
There's actually a few pallets that do it like this. But i'll try to write a test to be sure it even works as i expect. I think ive found an example of a migration test as well.

Reply:
@tonyalaribe https://github.com/paritytech/substrate/blob/aa/safely-destroy-large-assets/frame/nomination-pools/src/lib.rs#L1144L1148